pageserver/src/tenant/timeline/init.rs

```diff
@@ -0,0 +1,211 @@
+pub(super) fn scan_timeline_dir(path: &Path) -> anyhow::Result<Vec<Discovered>> {
+    let mut ret = Vec::new();
+
+    for direntry in std::fs::read_dir(path)? {
+        let direntry = direntry?;
+        let direntry_path = direntry.path();
+        let file_name = direntry.file_name();
+
+        let discovered = {
+            let fname = file_name.to_string_lossy();
+
+            match LayerFileName::from_str(&fname) {
+                Ok(LayerFileName::Image(file_name)) => {
+                    let file_size = direntry_path.metadata()?.len();
+                    Discovered::Layer(file_name.into(), file_size)
+                }
+                Ok(LayerFileName::Delta(file_name)) => {
+                    let file_size = direntry_path.metadata()?.len();
+                    Discovered::Layer(file_name.into(), file_size)
+                }
+                Err(_) => {
+                    if fname == METADATA_FILE_NAME {
+                        Discovered::Metadata
+                    } else if fname.ends_with(".old") {
+                        // ignore these
+                        Discovered::IgnoredBackup
+                    } else if remote_timeline_client::is_temp_download_file(&direntry_path) {
+                        Discovered::TemporaryDownload(file_name)
+                    } else if is_ephemeral_file(&fname) {
+                        Discovered::Ephemeral(file_name)
+                    } else if is_temporary(&direntry_path) {
+                        Discovered::Temporary(file_name)
+                    } else {
+                        Discovered::Unknown(file_name)
+                    }
+                }
+            }
+        };
+
+        ret.push(discovered);
+    }
+
+    Ok(ret)
+}
+
+/// Decision on what to do with a layer file after considering it's local and remote metadata.
+#[derive(Clone)]
+pub(super) enum Decision {
+    /// The layer is not present locally.
+    Evicted(LayerFileMetadata),
+    /// The layer is present locally, but local metadata does not match remote; we must
+    /// delete it and treat it as evicted.
+    UseRemote {
+        local: LayerFileMetadata,
+        remote: LayerFileMetadata,
+    },
+    /// The layer is present locally, and metadata matches.
+    UseLocal(LayerFileMetadata),
+    /// The layer is only known locally, it needs to be uploaded.
+    NeedsUpload(LayerFileMetadata),
+}
+
+/// The related layer is is in future compared to disk_consistent_lsn, it must not be loaded.
+#[derive(Debug)]
+pub(super) struct FutureLayer;
+
+/// Merges local discoveries and remote [`IndexPart`] to a collection of decisions.
+///
+/// This function should not gain additional reasons to fail than [`FutureLayer`], consider adding
+/// the checks earlier to [`scan_timeline_dir`].
+pub(super) fn reconcile(
+    discovered: Vec<(LayerFileName, u64)>,
+    index_part: Option<&IndexPart>,
+    disk_consistent_lsn: Lsn,
+) -> Vec<(LayerFileName, Result<Decision, FutureLayer>)> {
```

Comment:
Explain where disk_consistent_lsn comes from.

```suggestion
/// Merges local discoveries and remote [`IndexPart`] to a collection of decisions.
/// 
/// Assumes that `disk_consistent_lsn` is the result of [`crate::tenant::merge_local_and_remote_metadata`].
///
/// This function should not gain additional reasons to fail than [`FutureLayer`], consider adding
/// the checks earlier to [`scan_timeline_dir`].
pub(super) fn reconcile(
    discovered: Vec<(LayerFileName, u64)>,
    index_part: Option<&IndexPart>,
    disk_consistent_lsn: Lsn,
) -> Vec<(LayerFileName, Result<Decision, FutureLayer>)> {
```

Reply:
Here the disk_consistent_lsn is from Timeline. And that disk_consistent_lsn comes from `merge_local_and_remote_metadata`.

However, how is it relevant here? I am thinking we don't need to muddy the waters here; or grow that "boundary of knowledge" to here. This function works on all `disk_consistent_lsn` no matter where it came from. In code there's only single path as well.